### ISSUE TITLE
performance suffers vs. std

### ISSUE THREAD

**Issue by mikedilger** [alexcrichton/bufstream#1]
I'm running a web server with typical pages handled in <1ms.  Once rust-postgres moved to this package, my typical performance of pages accessing the database moved to 41ms (never less than 40ms).  By reverting to std version of this library in a fork of rust-postgres, performance returned.

downstream bug report: https://github.com/sfackler/rust-postgres/issues/123

**Comment by alexcrichton** [comment]
Do you have a benchmark that I could test this out on? Also, are you sure that you're building a `--release` binary as well?

**Comment by mikedilger** [comment]
Yes I built as `--release`.

You can use my webserver project to benchmark changes, but it's not small, and it does setup a postgres database (it's pretty slick about it though).  It's close-source; I added you to the private github so you can use it for now.  https://github.com/mikedilger/joist
Edit joist.toml before running it the first time.  It creates the database for you.

Fashioning something smaller and simpler to test it out is another option.

**Comment by alexcrichton** [comment]
Hm I seem to be having a bit of trouble getting a database up and running (I'm terrible at managing database permission), could you perhaps run `perf record` on the server binary and then hit it with a few requests, posing the results of `perf report` here?

**Comment by mikedilger** [comment]
updated comment above, just seeing yours, will do...

**Comment by mikedilger** [comment]
# To display the perf.data header info, please use --header/--header-only options.
#
# Samples: 62  of event 'cycles'
# Event count (approx.): 24852552
#
# Overhead  Command         Shared Object                                              Symbol
# ........  .......  ....................  ..................................................
#
    11.99%    joist  libc-2.20.so          [.] _dl_addr                                      
    11.32%    joist  libc-2.20.so          [.] _IO_un_link                                   
    11.31%    joist  libnss_files-2.20.so  [.] 0x00000000000037f4                            
    11.27%    joist  [kernel.kallsyms]     [k] __d_lookup_rcu                                
    10.84%    joist  joist                 [.] InnerConnection::connect::h6645464910445779678
    10.79%    joist  [kernel.kallsyms]     [k] __mutex_init                                  
     8.90%    joist  [kernel.kallsyms]     [k] kmem_cache_alloc_node_trace                   
     8.12%    joist  [kernel.kallsyms]     [k] __mod_zone_page_state                         
     1.98%    joist  [kernel.kallsyms]     [k] nf_conntrack_in                               
     1.93%    joist  [kernel.kallsyms]     [k] rb_insert_color                               
     1.90%    joist  [kernel.kallsyms]     [k] get_futex_key                                 
     1.83%    joist  [kernel.kallsyms]     [k] ipv6_defrag                                   
     1.68%    joist  [kernel.kallsyms]     [k] inet_recvmsg                                  
     1.60%    joist  [kernel.kallsyms]     [k] __dev_queue_xmit                              
     1.58%    joist  [kernel.kallsyms]     [k] tcp_transmit_skb                              
     1.30%    joist  [kernel.kallsyms]     [k] lru_add_drain_cpu                             
     1.16%    joist  [kernel.kallsyms]     [k] setup_arg_pages                               
     0.50%    joist  [kernel.kallsyms]     [k] intel_pmu_enable_all

**Comment by alexcrichton** [comment]
Hm that doesn't look like there's much related to bufstream in that profile... Maybe let's try another vector!

So the error I get when running the app is:

```
thread '<main>' panicked at 'Unable to initialize database: joist_pgsql/src/setup.rs:119  Failed to setup ownerconn: joist_pgsql/src/setup.rs:241  Unable to setup owner connection to database: : Database Connect Error DbError(DbError { severity: "FATAL", code: InvalidAuthorizationSpecification, message: "no pg_hba.conf entry for host \"127.0.0.1\", user \"joistdemo_owner\", database \"joistdemo\", SSL off", detail: None, hint: None, position: None, where_: None, schema: None, table: None, column: None, datatype: None, constraint: None, file: "auth.c", line: 496, routine: "ClientAuthentication" })', src/main.rs:140
```

Do you know how I would give `joistdemo_owner` access to the database? (I'm such a postgres newb)

**Comment by mikedilger** [comment]
You need to edit pg_hba.conf, depending on your distribution could be in `/etc/postgresql/` or `/var/lib/postgresql/` or `/var/lib/postgresql/{version}`, or similar.

Add a line like this (I think this will work):
`host    joistdemo    all    127.0.0.1    trust`

**Comment by mikedilger** [comment]
Also, I pushed another commit to benchmark_point, but the only relevant bit (based on your comments) would now be setting `threads` in `joist.toml` to be large (due to a hyper keep-alive issue that ties up threads)

**Comment by sfackler** [comment]
Isn't the final `get_mut` call here causing write and flush to bypass the buffer entirely? https://github.com/alexcrichton/bufstream/blob/master/src/lib.rs#L154

**Comment by sfackler** [comment]
Fixed in #2
